compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -119,6 +119,46 @@ mod sty;
 
 // Data types
 
+/// A pointer to a value that is known to be unique, e.g. via interning.
+/// The uniqueness property means that equality and hashing can be done on the
+/// pointer rather than the value's contents, which can improve performance.
+#[derive(Clone, Copy, Debug, HashStable)]
+pub struct Uniq<'tcx, T>(&'tcx T);
+
+impl<'tcx, T> Uniq<'tcx, T> {
+    /// Create a new `Uniq` value. This value *must* be known to be unique, and
+    /// it *must* remain unique in the future. If this condition is violated it
+    /// will cause incorrect behaviour, though it shouldn't cause undefined
+    /// behaviour. For this reason it is marked `unsafe`.
+    unsafe fn new(ptr: &'tcx T) -> Self {
```

Comment:
considering that getting this wrong is memory safe, i would prefer
```rust
fn new_unchecked(ptr: &'tcx T) -> Self {
```
here instead :laughing:

Reply:
In my stash I've written these methods, that should cover a lot of use cases, preventing `new_unchecked` callsites.
```rust
    pub unsafe fn new_unchecked(inner: &'tcx T) -> Self {
        Uniq(inner)
    }

    pub fn from_arena_interner(
        value: T,
        arena: &'tcx WorkerLocal<Arena<'tcx>>,
        interner: &ShardedHashMap<&'tcx T, ()>,
    ) -> Self {
        let inner = interner.intern(value, |v| arena.alloc(v));
        Uniq(inner)
    }

    pub fn from_make<K, F: FnOnce(K) -> T>(
        key: K,
        make: F,
        arena: &'tcx WorkerLocal<Arena<'tcx>>,
        interner: &ShardedHashMap<&'tcx T, ()>,
    ) -> Self {
        let inner = interner.intern(key, move |key| arena.alloc(make(key)));
        Uniq(inner)
    }
```